### PR TITLE
Add layerwise SPN PyTorch code

### PR DESCRIPTION
**Pull request by steven-lang**
Include the code for the PyTorch layerwise SPN implementation and a README file into `spn.algorithms.layerwise`.

### CODE REVIEW

**Comment by pranavsubramani** [comment]
Hey Steven, this commit looks very interesting and I'd like to take a closer look at it. What I'm curious about is the following (from a brief review of the code), I had built a layerwise pytorch implementation on spflow and it's up under src/gpu/PyTorch -- one major issue that I could not overcome was the fact that sparse backpropagation is currently not supported in pytorch which makes things a lot more memory intensive/slower. However, there is a massive speedup from making everything matrix multiplications.

I'll take a closer look into the structure of the layerwise implementation.

**Comment by steven-lang** [comment]
Hey Pranav, this implementation does not use any sparse tensors and doesn't suffer from the issue you described. If you have any questions feel free to ask!

File: src/spn/algorithms/layerwise/README.rst [hunk 1]
```diff
@@ -0,0 +1,193 @@
+============================================================
+torchspn - A PyTorch Layerwise Sum-Product Network Framework
+============================================================
+
```
Comment (pranavsubramani) [inline src/spn/algorithms/layerwise/README.rst]: Is there a reason to package this separately than spflow, while it's a good idea to disentangle the tensorflow and pytorch backends, ideally the glue sticking them together should be the spflow structure.

**Comment by pranavsubramani** [comment]
Wow, that's pretty awesome! I'll add some comments as we go along to try and understand some of the core logic (your documentation is extremely helpful for this!)

File: src/spn/algorithms/layerwise/README.rst (hunk 1 above)
Comment (steven-lang) [reply src/spn/algorithms/layerwise/README.rst]: Woops, I initially wanted to publish it as separate pypi package until it came to my mind that we could simply include it into SPFlow. I already had the readme written out so this was just an artifact. I've pushed an update to the readme.

Comment (pranavsubramani) [reply src/spn/algorithms/layerwise/README.rst]: Great! Thanks for the clarification!

File: src/spn/algorithms/layerwise/clipper.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+LOWER_BOUND = torch.tensor(1e-1)
+
+
+class DistributionClipper(object):
+    """
+    Clip distribution parameters.
+    """
+
+    def __init__(self, device, lower_bound=1e-5):
```
Comment (pranavsubramani) [inline src/spn/algorithms/layerwise/clipper.py]: Should we default the device argument as well? Is there a benefit to not doing it?

File: src/spn/algorithms/layerwise/distributions.py [hunk 3]
```diff
@@ -0,0 +1,465 @@
+"""
+Module that contains a set of distributions with learnable parameters.
+"""
+from abc import ABC
```
Comment (pranavsubramani) [inline src/spn/algorithms/layerwise/distributions.py]: What is ABC here?

File: src/spn/algorithms/layerwise/clipper.py (hunk 2 above)
Comment (steven-lang) [reply src/spn/algorithms/layerwise/clipper.py]: I suppose we could default to `torch.device("cpu")` as is the default in PyTorch for the tensor class. I'll update that with an additional `to(device)` method to make it more PyTorch-y

File: src/spn/algorithms/layerwise/distributions.py (hunk 3 above)
Comment (steven-lang) [reply src/spn/algorithms/layerwise/distributions.py]: Its Python's `AbstractBaseClass` to ensure proper inheritance for the `Leaf` class and its subclasses.

**Comment by steven-lang** [comment]
I've also added my implementation of RAT-SPNs with the layerwise pytorch code to the experiments

**Comment by AmurG** [comment]
Hey steven,

Just wanted to make a comment - the way you're doing product nodes essentially sums adjacent LL nodes, so 

1) Isn't this done more efficiently via conv with a kernel of all 1s, stride = kernel width ? 

2) To gain a larger generality, one can permute the variables before feeding them to the prod layer. Is this efficiently doable in pytorch ?

**Comment by steven-lang** [comment]
Hi AmurG, 

_Regarding 1)_
I previously had some experiments where I used `F.conv1d(...)` which forced me to permute the input twice and use the `groups` parameter. In the end it was slower in most cases (number of input features, channels etc) so I discarded it. But upon your comment I had another idea, namely to use `F.conv2d(...)` and treat dimensions differently such that no permutation is necessary and the filters do not have to be grouped. I've benchmarked the convolution vs no-convolution implementation (benchmark code see [here](https://gist.github.com/steven-lang/9d70fc912773a53cc3e1ce714d36919c)) and these are the results (run on a GTX 1080) (shortened, full numbers are in the gist):

```
conv= True
...
in_features=16384 , n_channels=100, n_children=2  , time=0.0078
in_features=16384 , n_channels=100, n_children=16 , time=0.0066
in_features=16384 , n_channels=100, n_children=256, time=0.0074
conv= False

... 
in_features=16384 , n_channels=100, n_children=2  , time=7.8873
in_features=16384 , n_channels=100, n_children=16 , time=0.9659
in_features=16384 , n_channels=100, n_children=256, time=0.7561
```

So we get a speed up factor between 10 and 1000, depending on the dimensions. This is **great**! Thanks for bringing me back to that idea :-)

_Regarding 2)_
I previously had a randomization parameter in product layers (default was `True`) but decided not to move the randomization into each product as it is a waste of computation time (indexing is slow) to do it in each product layer. As soon as the randomization took place once, there is no need for ruther shuffling. We could either let the user do the randomization on demand before he sends the data through the spn or we coudl provide a `Randomize` layer that does the job and can be put into the SPN wherever the user wants.
